sql: honour RLS policies during query-based backfill

spilchen opened the pull request:
Previously, schema changes that performed backfill operations from a query—such as those for materialized views (MQTs) or CREATE TABLE ... AS (CTAS)—executed the query as the node user. This user has administrative privileges and bypasses all Row-Level Security (RLS) policies, unintentionally exposing rows the originator of the change should not have been able to access.

This change ensures that such query-based backfills run under the privileges of the user who initiated the schema change. As a result, RLS policies are correctly enforced, and only the rows visible to the initiating user are included in the result.

Fixes #144816
Fixes #144776

Epic: CRDB-11724
Release note: none

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/144843)
<!-- Reviewable:end -->

spilchen committed:
sql: honour RLS policies during query-based backfill

Previously, schema changes that performed backfill operations from a query—such
as those for materialized views (MQTs) or CREATE TABLE ... AS (CTAS)—executed
the query as the node user. This user has administrative privileges and
bypasses all Row-Level Security (RLS) policies, unintentionally exposing rows
the originator of the change should not have been able to access.

This change ensures that such query-based backfills run under the privileges of
the user who initiated the schema change. As a result, RLS policies are
correctly enforced, and only the rows visible to the initiating user are
included in the result.

Fixes #144816
Fixes #144776

Epic: CRDB-11724
Release note: none

fqazi reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 2 of 2 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/144843#-:-OOT5oKM9OYQvHJadGF7:bnfp4nl)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @spilchen)



<!-- Sent from Reviewable.io -->

rafiss reviewed (approved):
lgtm! thanks for the nice tests

spilchen commented:
TFTR!

bors r+
